### PR TITLE
AML-3182 Initial Web Project

### PR DESCRIPTION
**Pull request by lazcool**
User Story: https://skillsfundingagency.atlassian.net/browse/AML-3182
Config PR: https://github.com/SkillsFundingAgency/das-employer-config/pull/187

Initial web project part 1. Adds a skeleton web project, including DI, bundling & minification, layout, url helper, initial config and ssl dev certificate.

Later PR's will need to finish the skeleton by adding authentication, error pages, google analytics, logging, App Insights, Anti Forgery...

Brings in layout from provider relationships, with these improvements and fixes:
* integrate GOV.UK Frontend using original sources, rather than build artifacts
* fix ESFA link in footer
* use exact GOV.UK Frontend colours (where colours are almost identical)
* use em for breakpoint units, so our own elements break consistently with those from GOV.UK Frontend~
* fix 'Home', 'Finance' line in menu splitting into 2 at certain widths when it shouldn't

~ Using `govuk-media-query` changes the media query unit to em from px.
There is debate as to which is the best to use, [this site](https://zellwk.com/blog/media-query-units/) claims Safari has issues using px for media queries, whereas [this site](https://adamwathan.me/dont-use-em-for-media-queries/) recommends using px.
However, if we use GOV.UK's mixin, at least the whole page will break consistently!

Note, [this page](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices) lists the browsers the site must look perfect in (the site should be functional in older browsers, but needn't look perfect).

Layout issues found (in provider relationships source), that we may or may not fix in this PR:
* when shrinking window the 'Home', 'Finance' line's left margin 
* line appears above 'Home', 'Finance' line in menu when hover
* top left GOV.UK underline looks odd
* your employer account line loses all left and right margin at some widths
* (on 'Home', 'Finance' line, main MA site uses 'More' drop-down, we could do the same)

### CODE REVIEW

File: src/SFA.DAS.EmployerFinance.Web/DependencyResolution/DefaultRegistry.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace SFA.DAS.EmployerFinance.Web.DependencyResolution
+{
+    public class DefaultRegistry : Registry
+    {
+        public DefaultRegistry()
+        {
+            For<IEmployerUrls>().Use<EmployerUrls>();
+
+//            Scan(s =>
```
Comment (elliotritchie) [inline src/SFA.DAS.EmployerFinance.Web/DependencyResolution/DefaultRegistry.cs]: Could the commented out code be removed?

File: src/SFA.DAS.EmployerFinance.Web/Filters/UrlsViewBagFilter.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace SFA.DAS.EmployerFinance.Web.Filters
+{
+    public class UrlsViewBagFilter : IAsyncActionFilter
+    {
+        public async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
+        {
+            var controller = context.Controller as Controller;
+            if (controller == null)
+                return;
```
Comment (elliotritchie) [inline src/SFA.DAS.EmployerFinance.Web/Filters/UrlsViewBagFilter.cs]: Could curly braces be added here to follow our code conventions?

File: src/SFA.DAS.EmployerFinance.Web/Filters/UrlsViewBagFilter.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+            // options: https://stackoverflow.com/questions/32459670/resolving-instances-with-asp-net-core-di
```
Comment (elliotritchie) [inline src/SFA.DAS.EmployerFinance.Web/Filters/UrlsViewBagFilter.cs]: Could the comment be removed?

File: src/SFA.DAS.EmployerFinance.Web/Startup.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        public IConfiguration Configuration { get; }
+
+        // This method gets called by the runtime. Use this method to add services to the container.
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.Configure<CookiePolicyOptions>(options =>
+            {
+                // This lambda determines whether user consent for non-essential cookies is needed for a given request.
```
Comment (elliotritchie) [inline src/SFA.DAS.EmployerFinance.Web/Startup.cs]: Could the comment be removed?

File: src/SFA.DAS.EmployerFinance.Web/Views/Shared/_Layout.cshtml [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        <link rel="stylesheet" href="~/css/das.css" />
+    </environment>
+    <environment exclude="Development">
+        @* todo once cdn includes version in name & a minified version, can pick this up from the cdn *@
+        <link rel="stylesheet" href="~/css/govuk-frontend.min.css" asp-append-version="true"/>
+        <link rel="stylesheet" href="~/css/das.min.css" asp-append-version="true"/>
+    </environment>
+</head>
+<body class="govuk-template__body">
+    @*todo @Html.Partial("_GoogleAnalyticsBody") *@
```
[the remark this answers is not in the document]
Comment (lazcool) [reply src/SFA.DAS.EmployerFinance.Web/Views/Shared/_Layout.cshtml]: Removed

File: src/SFA.DAS.EmployerFinance.Web/Views/Shared/_Layout.cshtml [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    <meta name="theme-color" content="#0b0c0c" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <meta property="og:image" content="/images/govuk-opengraph-image.png" />
+    <link rel="apple-touch-icon" sizes="180x180" href="/images/govuk-apple-touch-icon-180x180.png" />
+    <link rel="apple-touch-icon" sizes="167x167" href="/images/govuk-apple-touch-icon-167x167.png" />
+    <link rel="apple-touch-icon" sizes="152x152" href="/images/govuk-apple-touch-icon-152x152.png" />
+    <link rel="apple-touch-icon" href="/images/govuk-apple-touch-icon.png" />
+    <link rel="mask-icon" href="/images/govuk-mask-icon.svg" color="#0b0c0c" />
+    <link rel="shortcut icon" href="/images/favicon.ico" type="image/x-icon" />
+    @*todo @Html.Partial("_GoogleAnalyticsHead") *@
```
[the remark this answers is not in the document]
Comment (lazcool) [reply src/SFA.DAS.EmployerFinance.Web/Views/Shared/_Layout.cshtml]: Removed.

File: src/SFA.DAS.EmployerFinance.Web/Controllers/HomeController.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    public class HomeController : Controller
+    {
+        public IActionResult Index()
+        {
+            return View();
+        }
+
+        [Route("error")]
+        [ResponseCache(Duration = 0, Location = ResponseCacheLocation.None, NoStore = true)]
+        public IActionResult Error()
```
[the remark this answers is not in the document]
Comment (lazcool) [reply src/SFA.DAS.EmployerFinance.Web/Controllers/HomeController.cs]: I didn't touch the error handling in this PR. We can handle it in a later PR.

File: src/SFA.DAS.EmployerFinance.Web/Filters/UrlsViewBagFilter.cs (hunk 2 above)
Comment (lazcool) [reply src/SFA.DAS.EmployerFinance.Web/Filters/UrlsViewBagFilter.cs]: Done.

File: src/SFA.DAS.EmployerFinance.Web/Views/Shared/_Layout.cshtml [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+    <environment include="Development">
+        <link rel="stylesheet" href="~/css/govuk-frontend.css" />
+        <link rel="stylesheet" href="~/css/das.css" />
+    </environment>
+    <environment exclude="Development">
+        @* todo once cdn includes version in name & a minified version, can pick this up from the cdn *@
```
[the remark this answers is not in the document]
Comment (lazcool) [reply src/SFA.DAS.EmployerFinance.Web/Views/Shared/_Layout.cshtml]: Comment removed and developer notes markup file added to Solution Items.

File: src/SFA.DAS.EmployerFinance.Web/Views/Shared/_Layout.cshtml [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+                <div class="govuk-footer__meta-item">
+                    <a class="govuk-footer__link govuk-footer__copyright-logo" href="https://www.nationalarchives.gov.uk/information-management/re-using-public-sector-information/uk-government-licensing-framework/crown-copyright/">© Crown copyright</a>
+                </div>
+            </div>
+        </div>
+    </footer>
+
+<environment include="Development">
+    <script src="~/lib/jquery/jquery.js"></script>
+    @* todo split site out into 2 js files, and pick up all.js from the cdn? what about versioning? separate files will make it more likely to hit the 6 concurrent downloads limit and slow the site down, but serving all.js from cdn should be cheaper?*@
```
[the remark this answers is not in the document]
Comment (lazcool) [reply src/SFA.DAS.EmployerFinance.Web/Views/Shared/_Layout.cshtml]: Done.

File: src/SFA.DAS.EmployerFinance.Web/DependencyResolution/DefaultRegistry.cs (hunk 1 above)
Comment (lazcool) [reply src/SFA.DAS.EmployerFinance.Web/DependencyResolution/DefaultRegistry.cs]: Removed.

File: src/SFA.DAS.EmployerFinance.Web/Filters/UrlsViewBagFilter.cs (hunk 3 above)
Comment (lazcool) [reply src/SFA.DAS.EmployerFinance.Web/Filters/UrlsViewBagFilter.cs]: Removed & added to dev notes.

File: src/SFA.DAS.EmployerFinance.Web/Startup.cs (hunk 4 above)
Comment (lazcool) [reply src/SFA.DAS.EmployerFinance.Web/Startup.cs]: Removed.

File: README.md [hunk 10]
```diff
@@ -26,11 +26,67 @@
...
+> The two repositories above are private. If the links appear to be dead, make sure that you're logged into GitHub with an account that has access to these i.e. that you are part of the Skills Funding Agency Team organization.
 
+##### Add Certificates
 
+Execute DevInstall.ps1 as an admin in a legacy Powershell console (the script is currently not compatible with Powershell Core), to import required certificates into their appropriate store locations. If a dialog prompts you whether to install the dev certificate, click 'Yes'.
 
-### OS X Differences
+#### Development Tasks
+ 
+##### Update GOV.UK Frontend
```
[the remark this answers is not in the document]
Comment (lazcool) [reply README.md]: Nice 👍 !

Setup has been shortened to `npm install` and `copy-govuk-frontend-js.bat` has been replaced with a gulp task called 'js'.

I've set js and css pre-processing as the default gulp task, so they can be run using plain `gulp`. We can introduce a 'build' task if we introduce other tasks later, such as 'clean'.

I've also updated and expanded the readme markup file to match the new scheme.

**Pull request merged**